src/System.Runtime.InteropServices/tests/System/Runtime/InteropServices/Marshal/GetDelegateForFunctionPointerTests.cs

```diff
@@ -48,6 +48,7 @@ public void GetDelegateForFunctionPointer_CollectibleType_ReturnsExpected()
 
             Delegate functionDelegate = Marshal.GetDelegateForFunctionPointer(ptr, type);
             VerifyDelegate(functionDelegate, targetMethod);
+            GC.KeepAlive(type);
```

Comment:
This may be ok workaround that hides the bug, but it is a not a fix. This KeepAlive should not be required.

This is related to collectible assemblies. @janvorli may want to take a look at this once he gets back from vacation (mid August).

Reply:
@jkotas looking at the test code, it seems that the delegate `d` can be collected by GC. Then the `ptr` could point to garbage, the `COMDelegate::ConvertToDelegate` would call the UMEntryThunk::Decode and it would return a garbage too. The `COMDelegate::s_pDelegateToFPtrHash->LookupValue` would return `INVALIDENTRY` for the garbage and so we would check if the pMT->Collectible() returns true and throw the exception.

So my guess that the missing piece here is calling GC.Keepalive for the `d` instead of the `type`.